src/library/scala/concurrent/Future.scala

```diff
@@ -628,17 +696,20 @@ object Future {
    * @param futures   the `IterableOnce` of Futures in which to find the first completed
    * @return          the `Future` holding the result of the future that is first to be completed
    */
-  def firstCompletedOf[T](futures: IterableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] = {
-    val p = Promise[T]()
-    val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
-      override def apply(v1: Try[T]): Unit = getAndSet(null) match {
-        case null => ()
-        case some => some tryComplete v1
+  final def firstCompletedOf[T](futures: IterableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] =
+    if (futures.isEmpty) Future.never
+    else {
+      val p = Promise[T]()
+      val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
```

Comment:
Would it be possible to create an anonymous `Promise` that also extends from `Try[T] => Unit` and use `tryComplete`? I'd avoid the `AtomicReference`

Reply:
@fwbrasil The introduction of this AtomicReference was to address another issue (long before this PR).